Refactor causal mask generation and detection for nn.transformer

mikekgfb opened the pull request:
Summary:
* Move causal mask detection into a method _detect_causal_mask
* Add device and dtype parameters to generate_subsequent to avoid extra copies & conversions by passing directly to torch.full. 
* Create a private global-scope function _generate_subsequent because static class attribute member functions not supported by TorchScript resulting in torchscripting errors.
* Make TransformerEncoder and TransformerDecoder consistent w.r.t. is_causal handling by calling _detect_casual_mask

Test Plan: sandcastle & github 

Closing as the PR was re-exported as #105265 due to a tooling issue.

Differential Revision: D45827765

vadimkantorov commented on torch/nn/modules/transformer.py:
```diff
@@ -796,3 +804,41 @@ def _get_activation_fn(activation: str) -> Callable[[Tensor], Tensor]:
+
+        # Do not use `torch.equal` so we handle batched masks by
+        # broadcasting the comparison.
+        make_causal = bool((mask == causal_comparison).all())
```
does this cause any GPU syncs? or in this context it's okay?

mikekgfb replied:
I have not found a way around a GPU sync for this.  At present, this check is only performed in Transformer{En,De}coder and if no bool value for the hint is specified.  Because the value of the flag can be amortized over all layers when it's executed here, it is a net win.

The best results are obtained by setting is_causal to {True,False} and avoid this check and the concomitant sync altogether.

mikaylagawarecki commented on torch/nn/modules/transformer.py:
```diff
@@ -368,7 +374,7 @@ def __init__(self, decoder_layer, num_layers, norm=None):
     def forward(self, tgt: Tensor, memory: Tensor, tgt_mask: Optional[Tensor] = None,
                 memory_mask: Optional[Tensor] = None, tgt_key_padding_mask: Optional[Tensor] = None,
-                memory_key_padding_mask: Optional[Tensor] = None, tgt_is_causal: bool = False,
+                memory_key_padding_mask: Optional[Tensor] = None, tgt_is_causal: Optional[bool] = None,
```
For posterity, from offline discussion 

`bool` --> `Optional[bool]` was for consistency with `is_causal` in `TransformerEncoder` and is not BC-breaking since this kwarg was not in the 2.0 release


Separately, I have a question here -- why do we not do the same for `memory_is_causal`?

mikekgfb replied:
@janEbert submitted the original diff that added tgt_is_causal as a Boolean, this ratifies this change in TransformerDecoder further and aligns it with TransformerEncoder.

@janEbert do you have thoughts on the utility of adding the same causal Optional[bool] kwarg `memory_is_causal` ?

janEbert replied:
My thinking was that a user pretty much never supplies a causal memory mask and automatically checking for it would be a waste 99 % of the time. :)

mikekgfb replied:
We might ask whether we should introduce the kwarg memory_is_causal : Optional[bool], but then skip detect and just  `if memory_is_causal is None: memory_is_causal = False` with a comment to the effect that we're not checking because memory mask is not commonly a causal mask.

Or is that just forcing consistency in form on something that in practice doesn't have consistency and we should just add a sentence to the doc string why memory_mask has no corresponding is_causal?

mikaylagawarecki commented on torch/nn/modules/transformer.py:
```diff
@@ -826,3 +834,41 @@ def _get_activation_fn(activation: str) -> Callable[[Tensor], Tensor]:
+    if is_causal is None and mask is not None:
+        # Handles mask being 3-D or 4-D; since mask should be square,
+        # this returns the correct sequence length in any case.
+        sz = mask.size(-2)
```
Just so I understand, the initial path was written assuming a 2-D mask only, is that correct?

mikekgfb replied:
That's correct.  This recognizes the common case of 2D mask.  if somebody expanded that mask to 4 dimensions, we would not catch recognize such  causal mask.  We could allocate a 4D mask and compare against it as well.  However, that would require additional cycles and memory.  

That being said,  it is important that there is no "loss" in terms of correctness as`*is_causal` is a hint that bypasses mask usage if it is True, but uses the mask as specified when False.  As such, it's a performance optimization, and there's a benefit in recognizing the common case quickly.  I have yet to encounter a 4D-expanded causal mask, and even if we were to -- it seems that the occurrence of this is infinitesimally small.

If anybody has such a 4D causal mask, the easiest outcome might be for them to set `*is_causal=True` to indicate this. (The main feature of this is to upgrade all models written before the introduction of Accelerated Transformer that pass an attention mask to take advantage of the new Flash and memory efficient kernels.)

mikekgfb replied:
Update as per discussion.  We can pass in expected size for causal mask detection and only accept it as causal mask if it has the causal mask with right size for the provided input.

mikaylagawarecki commented on torch/nn/modules/transformer.py:
```diff
@@ -15,6 +15,21 @@
 
 __all__ = ['Transformer', 'TransformerEncoder', 'TransformerDecoder', 'TransformerEncoderLayer', 'TransformerDecoderLayer']
 
+def _generate_square_subsequent_mask(
```
Clarifying from the discussion offline, the reason for this helper is because we want to call this non-static variant in `_detect_is_causal_mask` so that `_detect_is_causal_mask` which is called by `TransformerEncoder.forward` / `TransformerDecoder.forward` can be torchscriptable is that correct?

mikekgfb replied:
That is correct!

mikaylagawarecki commented on torch/nn/modules/transformer.py:
```diff
@@ -826,3 +834,41 @@ def _get_activation_fn(activation: str) -> Callable[[Tensor], Tensor]:
+       may lead to incorrect and unpredictable execution - in some scenarios,
+       a causal mask may be applied based on the hint, in other execution
+       scenarios the specified mask may be used.  The choice may not appear
+       to be detrministic, in that a number of factors like alignment,
```
nit:
```suggestion
       to be deterministic, in that a number of factors like alignment,
```

mikekgfb replied:
Thank you!

mikaylagawarecki reviewed (commented):
High level question:

Seems like `tgt_is_causal` for `TransformerDecoder` was missing the autodetect logic that it was documented to have but this PR adds it is that correct? 

Separately, why do we not have this autodetect behavior for `memory_is_causal`?

mikaylagawarecki commented on test/test_nn.py:
```diff
@@ -3931,7 +3931,7 @@ def update_shape(shape, dim, new_dim_size):
         decoder_input_shape = correct_decoder_input_shape
         wrong_tgt_mask_size = tgt_len + 1
-        test(encoder_input_shape, decoder_input_shape, tgt_mask_len=wrong_tgt_mask_size, raises=True)
+        test(encoder_input_shape, decoder_input_shape, tgt_mask_len=wrong_tgt_mask_size)
```
why are we ok with no longer raising for wrong `tgt_mask`s?

janEbert replied:
My last state (which I must say is probably quite out of date by now) was also that we could leave all of the `raises` arguments as they were.

mikekgfb replied:
There used to be an else-branch on the second if

janEbert commented:
See also #98327.

janEbert commented on torch/nn/functional.py:
```diff
@@ -5261,12 +5261,13 @@ def multi_head_attention_forward(
-            "You may use the Transformer module method "
-            "`generate_square_subsequent_mask` to create this mask."
-        )
+    if is_causal:
+        if attn_mask is None:
```
Just out of interest, is there a specific reason (e.g. TorchScript) to separate the if-query?

mikekgfb replied:
There used to be an else-branch on the second if

janEbert replied:
In that case I think it would make sense to revert the change, it's a bit nicer to look at with the combined if-query in my opinion.

Commit:
Refactor causal mask generation and detection for nn.transformer (#101487)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/101487

* Move causal mask detection into a method _detect_causal_mask
* Add device and dtype parameters to generate_subsequent to avoid extra copies & conversions by passing directly to torch.full
* Make TransformerEncoder and TransformerDecoder consistent w.r.t. is_causal handling by calling _detect_casual_mask

Test Plan: sandcastle

Differential Revision: D45827765

fbshipit-source-id: d3e40ba53ef7d03d631aca815c0be00aae0e95f2

izaitsevfb commented:
Closing as the PR was re-exported as #105154

mikekgfb commented:
Continued in #105265
